partitura/utils/music.py

```diff
@@ -2871,6 +2871,116 @@ def get_matched_notes(spart_note_array, ppart_note_array, alignment):
     return np.array(matched_idxs)
 
 
+def slice_ppart_by_time(
+    ppart, 
+    start_time: float, 
+    end_time: float,
+    clip_note_off: bool=True,
+    reindex_notes: bool=True):
+    """
+    Get a slice of a PeformedPart by time
+
+    Parameters
+    ----------
+    ppart : `PerformedPart` object
+    start_time : float
+        Starting time in seconds 
+    end_time : float
+        End time in seconds
+    clip_note_off : bool
+        Clip note_off time to end_time
+    reindex_notes : bool
+        Reindex notes in slice starting from n0
+
+    Returns
+    -------
+    ppart_slice :  `PerformedPart` object 
+        A copy of input ppart containing notes, programs and control information 
+        only between `start_time` and `end_time` of ppart
+    """
+    from partitura.performance import PerformedPart
+    from copy import deepcopy
+
+    if not isinstance(ppart, PerformedPart):
+        raise ValueError("Input is not an instance of PerformedPart!")
+
+    assert(start_time < end_time), "Start time not less than end time!"
+
+    ppart_slice = deepcopy(ppart)
+    
+    # get ppq if PerformedPart contains it, 
+    # else skip time_tick info when e.g. created with 'load_performance_midi'
+    try:
+        ppq = ppart.ppq
+    except AttributeError:
+        ppq = None
+
+    controls_slice = []
+    for cc in ppart.controls:
+        if cc['time'] >= start_time and cc['time'] <= end_time:
+            new_cc = cc.copy()
+            new_cc['time'] -= start_time
+            if ppq: 
+                new_cc['time_tick'] = int(2 * ppq * cc['time'])
+            controls_slice.append(new_cc)
+
+    programs_slice = []
+    for pr in ppart.programs:
+        if pr['time'] >= start_time and pr['time'] <= end_time:
+            new_pr = pr.copy()
+            new_pr['time'] -= start_time
+            if ppq:
+                new_pr['time_tick'] = int(2 * ppq * pr['time'])
+            programs_slice.append(new_pr)
+
+    notes_slice = []
+    note_id = 0
+    for note in ppart.notes: 
```

Comment:
maybe it is better here to loop through `performance.note_array()` this is running in numpy so it can be done faster. 
I would propose using numpy functions such as:
```
import numpy as np
import partitura as pt
idx = np.where((note_array["onset_sec"] >= slice_start) & ((note_array["onset_sec"]+note_array["duration_sec"] <= slice_end))
new_note_array = note_array[idx]
ppart_slice = pt.performance.PerformedPart.from_note_array(new_note_array)
```

Reply:
That was my initial idea, but unfortunately this proposed method would not work due to the loss of information when converting to `performance.note_array()` as there is no distinction between 'note_off' and 'sound_off'. This creates a new performed part with notes list which contain different 'note_off" values than the original ones.
Therefore I would propose sacrificing efficiency in this case in order to obtain accuracy.

Reply:
Maybe it would be interesting to have an optional inclusion of such parameters to the note array then, i.e.
`performance.note_array(include_sound_off=True)`.  I propose to move this to a discussion so we can decide how to add functionality to the package that will be more inclusive for future potential uses.